test/unit/math/opencl/matrix_cl_test.cpp

```diff
@@ -26,4 +27,23 @@ TEST(MathMatrixCL, matrix_cl_types_creation) {
   test_matrix_creation<long double>();
 }
 
+#ifdef STAN_OPENCL_CACHE
+TEST(MathMatrixCL, matrix_cl_cache) {
+  using stan::math::matrix_cl;
+  Eigen::MatrixXd m(2, 2);
+  m << 1, 2, 3, 4;
+
+  const matrix_cl<double> m1_cl = matrix_cl<double>::constant(m);
+
+  cl_mem mem_handle = m.opencl_buffer_();
+  EXPECT_EQ(mem_handle, m1_cl.buffer()());
+
+  const matrix_cl<double> m2_cl = matrix_cl<double>::constant(m);
+
+  EXPECT_EQ(mem_handle, m.opencl_buffer_());
+  EXPECT_EQ(mem_handle, m1_cl.buffer()());
+  EXPECT_EQ(mem_handle, m2_cl.buffer()());
```

Comment:
These are great. I would add 2 more tests: 
1. If we are going to stick with the `CACHE` flag or use a `NOCACHE` flag there should be a separate test to check the behavior when its not set. Again, I dont think the flag is actually needed.

2. a test that checks that no caching happens when using a regular constructor from eigen.